doc/cueprint.1
.TH "cueprint" "1"
.SH NAME
cueprint \- report disc and track information from a CUE or TOC file
.SH SYNOPSIS
.B cueprint
[ {
.B \-i
.I format
|
.BR \-\-input\-format =\fIformat\fP
} {
.B \-n
.I number
|
.BR \-\-track\-number =\fInumber\fP
} {
.B \-d
.I template
|
.BR \-\-disc\-template =\fItemplate\fP
} {
.B \-t
.I template
|
.BR \-\-track\-template =\fItemplate\fP
} ]
[
.I file
\&... ]
.br
.B cueprint \-h | \-\-help
.br
.B cueprint \-V | \-\-version
.SH DESCRIPTION
.B cueprint
reports disc and track information from CUE and TOC files, which are
typically created by compact disc ripping software.
.PP
By default,
.B cueprint
reports a default set of disc information, and a default set of information
for each track on the disc.
With options, the report can be restricted to a specific track, and the
presentation of the disc and track information can be extensively
customized using
.BR printf (3)-style
format strings (referred to here as \(oqtemplates\(cq to avoid confusion
with the concept of file format).
All characters in a template are reproduced as-is in the output except for
conversions (which begin with
.RB \(oq % \(cq)
and escapes (which begin with
.RB \(oq \e \(cq).
.PP
If no filenames are specified,
.B cueprint
reads from standard input, and an input format option
.I must
be specified.
If one or more filenames is provided, but the input format option is not
specified, the input format will be guessed based on each file's suffix
(e.g.,
.I .cue
or
.IR .toc ).
This heuristic is case-insensitive.
.SS Conversions
A conversion has the form
.RB \(oq % [ \fIflags\fP ][ \fIwidth\fP ][ .\fIprecision\fP ] \fItype\fP \(cq.
.PP
.I flags
may be zero or more of the following:
.TS
nokeep;
l	l.
Character	Meaning
_
\-	left-justify expansion
+	place sign before numbers
\(oq \(cq (space)	place a blank space before a positive number
0	pad numbers with zeroes
.TE
.\"#	(unused)
.PP
.I width
is the minimum field width.
.I precision
is the maximum width for strings.
.I type
is a single character which specifies the conversion type \(em apart from
.BR % ,
it is the only mandatory part of the conversion.
.PP
The available conversion types are presented in the table below; disc
conversion types are presented in the left half of the table, and track
conversion types in the right half.
Disc template expansion characters are valid for both disc and track
templates.
.TS
nokeep;
l	l	l	l.
Character	Conversion	Character	Conversion
_
A	album arranger	a	track arranger
C	album composer	c	track composer
G	album genre	g	track genre
		i	track ISRC
M	album message	m	track message
N	number of tracks	n	track number
P	album performer	p	track performer
S	album songwriter
T	album title	t	track title
U	album UPC/EAN	u	track ISRC (CD-TEXT)
.TE
.PP
Any other character used as a conversion type expands to itself.
This is how a literal percent sign is placed in the template; i.e.,
.RB \(oq %% \(cq
expands to
.RB \(oq % \(cq.
.SS Escapes
The recognized escapes are all single characters, and listed in the table
below.
.TS
nokeep;
l	l.
Escape Sequence	Expansion
_
\ea	alert (bell)
\eb	backspace
\ef	formfeed
\en	newline
\er	carriage return
\et	horizontal tab
\ev	vertical tab
\e0	null
.TE
.PP
Any other character used after the
.RB \(oq \e \(cq
in an escape sequence expands to itself.
This is how a literal escape character is placed in the template; i.e.,
.RB \(oq \e\e \(cq
expands to
.RB \(oq \e \(cq.
.SH OPTIONS
.TP
.BR \-d " \fItemplate\fP, " \-\-disc\-template=\fItemplate\fP
set disc template (see
.B Conversions
)
.TP
.BR \-h ", " \-\-help
displays a usage message and exits.
.TP
.BR \-i " \fIformat\fP, " \-\-input\-format=\fIformat\fP
sets the expected format of the input file(s) to
.IR format ,
which must be either
.B cue
or
.BR toc .
.TP
.BR \-n " \fInumber\fP, " \-\-track\-number=\fInumber\fP
only print track information for a single track.
The default is to print information for all tracks.
.TP
.BR \-t " \fItemplate\fP, " \-\-track\-template=\fItemplate\fP
set track template (see
.B Conversions
)
.TP
.B \-V ", " \-\-version
displays version information and exits.
.SH "EXIT STATUS"
.B cueprint
exits with status zero if it successfully reports information from each
input file, and nonzero if there were problems.
.SH EXAMPLES
To display disc and track information (using the default template for
both):
.PP
.RB "% " "cueprint album.cue"
.PP
To print the number of tracks in a CUE file:
.PP
.RB "% " "cueprint -d \(aq%N\en\(aq album.cue"
.SH AUTHOR
Cuetools was written by Svend Sorensen.
Branden Robinson contributed fixes and enhancements to the utilities and
documentation.
.SH "SEE ALSO"
.BR cuebreakpoints (1),
.BR cueconvert (1),
.BR printf(3)


NEWS
Changes since 1.3

* Applied most of Branden Robinson's Debian patch
  (cuetools_1.3-2.diff.gz). This included many fixes and enhancements
  to tools and manual pages. cuebreakpoints and cueconvert will exit if
  they fail on any of the input files. cuebreakpoints, cueconvert, and
  cuetools will exit with a nonzero return value if there was a failure
  on an input file.

* Added version option to all tools.

* Single quoted string support has been removed.  This was broken, and I am not
  sure of any software that supports single quoted strings in cue or toc files.

Changes since 1.2

* Build system has been rewritten using automake/autoconf.

* Added append, prepend, and split pregap modes to cuebreakpoints.

* Fixed possible compilation error with Bison.

* All tools accept long options.

* Programs exit if --input-format or --output-format is an illegal value.

Changes since 1.1

* cuebreakpoints was not printing the last track breakpoint.  This has been
  fixed.

* cueprint now prints a zero-length string for undefined values.  Flags, width,
  and precision are applied to conversion characters which have no conversion.

* Added a tagging script, cuetag.sh, which uses cueprint, vorbiscomment, and
  mp3info to tag Ogg Vorbis and MP3 files.  This needs more testing and
  documentation.

* Added a track selection flag to cueprint.

Changes since 1.0

* File formats documentation has been readded and updated (docs/formats.txt).

* ``CDTEXTFILE`` is now a recognized (but ignored) keyword.

* If only a disc or track template is specified on the command line, then only
  the disc or track information is printed, respectively.

* cueprint now supports flags and width and precision values for template
  expansion (like the printf command).

Changes since 0.6

* cuetools-1.x is a significant code rewrite of 0.x.  The cue/toc parsing code
  has been reimplemented in lex/yacc.

* cuerename has been replaced by the more general utility cueprint.

* Usage syntax of the utilities has changed.  See the respective man pages.


src/tools/cueprint.c
/*
 * cueprint.c -- print cd information based on a template
 *
 * Copyright (C) 2004, 2005, 2006, 2007 Svend Sorensen
 * For license terms, see the file COPYING in this distribution.
 */

#include <ctype.h>	/* isdigit() */
#include <getopt.h>	/* getopt_long() */
#include <stdio.h>	/* fprintf(), printf(), snprintf(), stderr */
#include <stdlib.h>	/* exit() */
#include <string.h>	/* strcasecmp() */
#include "cuefile.h"

#if HAVE_CONFIG_H
#include "config.h"
#else /* not HAVE_CONFIG_H */
#define PACKAGE_STRING "cueprint"
#endif /* HAVE_CONFIG_H */

/* default templates */

#define D_TEMPLATE "\
Disc Information\n\
arranger:	%A\n\
composer:	%C\n\
genre:		%G\n\
message:	%M\n\
no. of tracks:	%N\n\
performer:	%P\n\
songwriter:	%S\n\
title:		%T\n\
UPC/EAN:	%U\n\
"

#define T_TEMPLATE "\
\n\
Track %n Information\n\
arranger:	%a\n\
composer:	%c\n\
genre:		%g\n\
ISRC:		%i\n\
message:	%m\n\
track number:	%n\n\
perfomer:	%p\n\
title:		%t\n\
ISRC (CD-TEXT):	%u\n\
"

/* default string to print for unset (NULL) values */
#define VALUE_UNSET ""

/*
 * *_get_* functions can return an int or a char *
 */
typedef union {
	int ival;
	char *sval;
	char cval;
} Value;

char *progname;

/* Print usage information and exit */
void usage(int status)
{
	if (0 == status) {
		printf("Usage: %s [option...] [file...]\n", progname);
		printf("Report disc and track information from a CUE or TOC file.\n"
		       "\n"
		       "OPTIONS\n"
		       "-h, --help 			print usage\n"
		       "-i, --input-format cue|toc	set format of file(s)\n"
		       "-n, --track-number <number>	only print track information for single track\n"
		       "-d, --disc-template <template>	set disc template\n"
		       "-t, --track-template <template>	set track template\n"
		       "-V, --version			print version information\n"
		       "\n"
		       "Default disc template: %s\n"
		       "Default track template: %s\n"
		       "See the manual page for more information.\n",
		       D_TEMPLATE, T_TEMPLATE);
	} else {
		fprintf(stderr, "Try `%s --help' for more information.\n", progname);
	}

	exit (status);
}

/* Print version information and exit */
void version()
{
	printf("%s\n", PACKAGE_STRING);

	exit(0);
}

/*
 * TODO: Shouldn't we be using vprintf() to help us out with this stuff?
 * (Branden Robinson)
 */
void disc_field(char *conv, int length, Cd *cd, Value *value)
{
	char *c;	/* pointer to conversion character */

	Cdtext *cdtext = NULL;
	cdtext = cd_get_cdtext(cd);

	c = conv + length - 1;

	/*
	 * After setting value, set *c to specify value type:
	 * 'd' integer
	 * 's' string
	 * 'c' character
	 */
	switch (*c) {
	case 'A':
		value->sval = cdtext_get(PTI_ARRANGER, cdtext);
		*c = 's';
		break;
	case 'C':
		value->sval = cdtext_get(PTI_COMPOSER, cdtext);
		*c = 's';
		break;
	case 'G':
		value->sval = cdtext_get(PTI_GENRE, cdtext);
		*c = 's';
		break;
	case 'M':
		value->sval = cdtext_get(PTI_MESSAGE, cdtext);
		*c = 's';
		break;
	case 'N':
		value->ival = cd_get_ntrack(cd);
		*c = 'd';
		break;
	case 'P':
		value->sval = cdtext_get(PTI_PERFORMER, cdtext);
		*c = 's';
		break;
	case 'R':
		value->sval = cdtext_get(PTI_ARRANGER, cdtext);
		*c = 's';
		break;
	case 'S':
		value->sval = cdtext_get(PTI_SONGWRITER, cdtext);
		*c = 's';
		break;
	case 'T':
		value->sval = cdtext_get(PTI_TITLE, cdtext);
		*c = 's';
		break;
	case 'U':
		value->sval = cdtext_get(PTI_UPC_ISRC, cdtext);
		*c = 's';
		break;
	default:
		value->cval = *c;
		*c = 'c';
		break;
	}
}

void track_field(char *conv, int length, Cd *cd, int trackno, Value *value)
{
	char *c;	/* pointer to conversion character */

	Track *track = NULL;
	Cdtext *cdtext = NULL;

	track = cd_get_track(cd, trackno);
	cdtext = track_get_cdtext(track);

	c = conv + length - 1;

	switch (*c) {
	case 'a':
		value->sval = cdtext_get(PTI_ARRANGER, cdtext);
		*c = 's';
		break;
	case 'c':
		value->sval = cdtext_get(PTI_COMPOSER, cdtext);
		*c = 's';
		break;
	case 'f':
		value->sval = track_get_filename(track);
		*c = 's';
		break;
	case 'g':
		value->sval = cdtext_get(PTI_GENRE, cdtext);
		*c = 's';
		break;
	case 'i':
		value->sval = track_get_isrc(track);
		*c = 's';
		break;
	case 'm':
		value->sval = cdtext_get(PTI_MESSAGE, cdtext);
		*c = 's';
		break;
	case 'n':
		value->ival = trackno;
		*c = 'd';
		break;
	case 'p':
		value->sval = cdtext_get(PTI_PERFORMER, cdtext);
		*c = 's';
		break;
	case 's':
		value->sval = cdtext_get(PTI_SONGWRITER, cdtext);
		*c = 's';
		break;
	case 't':
		value->sval = cdtext_get(PTI_TITLE, cdtext);
		*c = 's';
		break;
	case 'u':
		value->sval = cdtext_get(PTI_UPC_ISRC, cdtext);
		*c = 's';
		break;
	default:
		disc_field(conv, length, cd, value);
		break;
	}

}

/*
 * Print a conversion specification.
 * [flag(s)][width][.precision]<conversion-char>
 */
void print_conv(char *start, int length, Cd *cd, int trackno)
{
	char *conv;	/* copy of conversion specification */
	Value value;
	char *c;	/* pointer to conversion-char */

	/* TODO: use strndup? */
	conv = malloc((unsigned) (length + 1));
	strncpy(conv, start, length);
	conv[length] = '\0';

	/* conversion character */
	if (0 == trackno) {
		disc_field(conv, length, cd, &value);
	} else {
		track_field(conv, length, cd, trackno, &value);
	}

	c = conv + length - 1;

	switch (*c) {
	case 'c':
		printf(conv, value.cval);
		break;
	case 'd':
		printf(conv, value.ival);
		break;
	case 's':
		if (NULL == value.sval)
			printf(conv, VALUE_UNSET);
		else
			printf(conv, value.sval);
		break;
	default:
		printf("%zu: ", strlen(conv));
		printf("%s", conv);
	}

	free(conv);
}

void cd_printf(char *format, Cd *cd, int trackno)
{
	char *c;	/* pointer into format */
	char *conv_start;
	int conv_length;

	for (c = format; '\0' != *c; c++) {
		if ('%' == *c) {
			conv_start = c;
			conv_length = 1;
			c++;

			/* flags */
			while ( \
				'-' == *c \
				|| '+' == *c \
				|| ' ' == *c \
				|| '0' == *c \
				|| '#' == *c \
			) {
				conv_length++;
				c++;
			}

			/* field width */
			/* '*' not recognized */
			while (0 != isdigit(*c)) {
				conv_length++;
				c++;
			}
			
			/* precision */
			/* '*' not recognized */
			if ('.' == *c) {
				conv_length++;
				c++;

				while (0 != isdigit(*c)) {
					conv_length++;
					c++;
				}
			}

			/* length modifier (h, l, or L) */
			/* not recognized */

			/* conversion character */
			conv_length++;

			print_conv(conv_start, conv_length, cd, trackno);
		} else {
			putchar(*c);
		}
	}
}

int info(char *name, int format, int trackno, char *d_template, char *t_template)
{
	Cd *cd = NULL;
	int ntrack;

	if (NULL == (cd = cf_parse(name, &format))) {
		fprintf(stderr, "%s: error: unable to parse input file"
		        " `%s'\n", progname, name);
		return -1;
	}

	ntrack = cd_get_ntrack(cd);

	if (-1 == trackno) {
		cd_printf(d_template, cd, 0);

		for (trackno = 1; trackno <= ntrack; trackno++) {
			cd_printf(t_template, cd, trackno);
		}
	} else if (0 == trackno) {
		cd_printf(d_template, cd, trackno);
	} else if (0 < trackno && ntrack >= trackno) {
		cd_printf(t_template, cd, trackno);
	} else {
		fprintf(stderr, "%s: error: track number out of range\n", progname);
		return -1;
	}

	return 0;
}

/* 
 * Translate escape sequences in a string.
 * The string is overwritten and terminated.
 * TODO: this does not handle octal and hexidecimal escapes
 *       except for \0
 */
void translate_escapes(char *s)
{
	char *read;
	char *write;

	read = s;
	write = s;

	while ('\0' != *read) {
		if ('\\' == *read) {
			read++;

			switch (*read) {
			case 'a':
				*write = '\a';
				break;
			case 'b':
				*write = '\b';
				break;
			case 'f':
				*write = '\f';
				break;
			case 'n':
				*write = '\n';
				break;
			case 'r':
				*write = '\r';
				break;
			case 't':
				*write = '\t';
				break;
			case 'v':
				*write = '\v';
				break;
			case '0':
				*write = '\0';
				break;
			default:
				/* ?, ', " are handled by the default */
				*write = *read;
				break;
			}
		} else {
			*write = *read;
		}

		read++;
		write++;
	}

	*write = '\0';
}

int main(int argc, char *argv[])
{
	int format = UNKNOWN;
	int trackno = -1;		/* track number (-1 = unspecified,
					                  0 = disc info) */
	char *d_template = NULL;	/* disc template */
	char *t_template = NULL;	/* track template */
	int ret = 0;			/* return value of info() */

	/* option variables */
	int c;
	/* getopt_long() variables */
	extern char *optarg;
	extern int optind;

	static struct option longopts[] = {
		{"help", no_argument, NULL, 'h'},
		{"input-format", required_argument, NULL, 'i'},
		{"track-number", required_argument, NULL, 'n'},
		{"disc-template", required_argument, NULL, 'd'},
		{"track-template", required_argument, NULL, 't'},
		{"version", no_argument, NULL, 'V'},
		{NULL, 0, NULL, 0}
	};

	progname = argv[0];

	while (-1 != (c = getopt_long(argc, argv, "hi:n:d:t:V", longopts, NULL))) {
		switch (c) {
		case 'h':
			usage(0);
			break;
		case 'i':
			if (0 == strcmp("cue", optarg)) {
				format = CUE;
			} else if (0 == strcmp("toc", optarg)) {
				format = TOC;
			} else {
				fprintf(stderr, "%s: error: unknown input file"
				        " format `%s'\n", progname, optarg);
				usage(1);
			}
			break;
		case 'n':
			trackno = atoi(optarg);
			break;
		case 'd':
			d_template = optarg;
			break;
		case 't':
			t_template = optarg;
			break;
		case 'V':
			version();
			break;
		default:
			usage(1);
			break;
		}
	}

	/* If no disc or track template is set, use the defaults for both. */
	/* TODO: alternative to strdup to get variable strings? */
	if (NULL == d_template && NULL == t_template) {
		d_template = strdup(D_TEMPLATE);
		t_template = strdup(T_TEMPLATE);
	} else {
		if (NULL == d_template) {
			d_template = strdup("");
		}

		if (NULL == t_template) {
			t_template = strdup("");
		}
	}

	/* Translate escape sequences. */
	translate_escapes(d_template);
	translate_escapes(t_template);

	/* What we do depends on the number of operands. */
	if (optind == argc) {
		/* No operands: report information about stdin. */
		ret = info("-", format, trackno, d_template, t_template);
	} else {
		/* Report information about each operand. */
		for (; optind < argc; optind++) {
			ret = info(argv[optind], format, trackno, d_template, t_template);
			/* Exit if info() returns nonzero. */
			if (!ret) {
				break;
			}
		}
	}

	return ret;
}
